sdk/python/cmd/pulumi-language-python/main.go

```diff
@@ -1009,3 +1017,154 @@ func (host *pythonLanguageHost) RunPlugin(
 
 	return closer.Close()
 }
+
+func (host *pythonLanguageHost) GenerateProject(
+	ctx context.Context, req *pulumirpc.GenerateProjectRequest,
+) (*pulumirpc.GenerateProjectResponse, error) {
+	loader, err := schema.NewLoaderClient(req.LoaderTarget)
+	if err != nil {
+		return nil, err
+	}
+
+	extraOptions := make([]pcl.BindOption, 0)
+	if !req.Strict {
+		extraOptions = append(extraOptions, pcl.NonStrictBindOptions()...)
+	}
+
+	// for python, prefer output-versioned invokes
+	extraOptions = append(extraOptions, pcl.PreferOutputVersionedInvokes)
+
+	program, diags, err := pcl.BindDirectory(req.SourceDirectory, loader, extraOptions...)
+	if err != nil {
+		return nil, err
+	}
+	if diags.HasErrors() {
+		rpcDiagnostics := make([]*codegenrpc.Diagnostic, 0)
+		for _, diag := range diags {
+			rpcDiagnostics = append(rpcDiagnostics, plugin.HclDiagnosticToRPCDiagnostic(diag))
+		}
```

Comment:
The final length of rpcDiagnostics is known. We can preallocate.
```suggestion
		rpcDiagnostics := make([]*codegenrpc.Diagnostic, len(diags))
		for i, diag := range diags {
			rpcDiagnostics[i] = plugin.HclDiagnosticToRPCDiagnostic(diag)
		}
```